kvbc/src/categorization/block_merkle_category.cpp

```diff
@@ -0,0 +1,398 @@
+                                                std::vector<std::optional<BlockId>>& versions) const {
+  auto hashed_keys = hashedKeys(keys);
+  multiGetLatestVersion(hashed_keys, versions);
+}
+
+void BlockMerkleCategory::multiGetLatestVersion(const std::vector<Hash>& hashed_keys,
+                                                std::vector<std::optional<BlockId>>& versions) const {
+  auto slices = std::vector<::rocksdb::PinnableSlice>{};
+  slices.reserve(hashed_keys.size());
+  auto statuses = std::vector<::rocksdb::Status>{};
+  statuses.reserve(hashed_keys.size());
+
+  db_->multiGet(BLOCK_MERKLE_LATEST_KEY_VERSION, hashed_keys, slices, statuses);
+  versions.clear();
+  for (auto i = 0ull; i < slices.size(); ++i) {
+    const auto& status = statuses[i];
+    const auto& slice = slices[i];
+    if (status.ok()) {
+      auto version = LatestKeyVersion{};
+      deserialize(slice, version);
+      if (version.block_id == 0) {
+        // 0 is a tombstone sentinel
+        versions.push_back(std::nullopt);
+      } else {
+        versions.push_back(version.block_id);
+      }
+    } else if (status.IsNotFound()) {
+      versions.push_back(std::nullopt);
+    } else {
+      throw std::runtime_error{"BlockMerkleCategory multiGet() failure: " + status.ToString()};
+    }
+  }
+}
+
+void BlockMerkleCategory::multiGetLatest(const std::vector<std::string>& keys,
+                                         std::vector<std::optional<MerkleValue>>& values) const {
+  auto hashed_keys = hashedKeys(keys);
+  std::vector<std::optional<BlockId>> versions;
+  versions.reserve(keys.size());
+  multiGetLatestVersion(hashed_keys, versions);
+
+  // Generate the set of versioned keys for all keys that have latest versions
+  auto versioned_keys = std::vector<Buffer>{};
+  auto found_versions = std::vector<BlockId>{};
+  for (auto i = 0u; i < hashed_keys.size(); i++) {
+    if (versions[i]) {
+      found_versions.push_back(*versions[i]);
+      versioned_keys.push_back(serialize(VersionedKey{KeyHash{hashed_keys[i]}, *versions[i]}));
+    }
+  }
+
+  values.clear();
+  // Optimize for all keys existing (having latest versions)
+  if (versioned_keys.size() == hashed_keys.size()) {
+    return multiGet(versioned_keys, found_versions, values);
+  }
+
+  // Retrieve only the keys that have latest versions
+  auto retrieved_values = std::vector<std::optional<MerkleValue>>{};
+  retrieved_values.reserve(versioned_keys.size());
+  multiGet(versioned_keys, found_versions, retrieved_values);
+
+  // Merge any keys that didn't have latest versions along with the retrieved keys.
+  auto value_index = 0u;
+  for (auto& version : versions) {
+    if (version) {
+      values.push_back(retrieved_values[value_index]);
+      ++value_index;
+    } else {
+      values.push_back(std::nullopt);
+    }
+  }
+  ConcordAssertEQ(values.size(), keys.size());
+}
+
+void BlockMerkleCategory::putKeys(NativeWriteBatch& batch,
+                                  uint64_t block_id,
```

Comment:
No treatment to the stale updates since deletion is not in place yet?

Reply:
I don't believe we have to store them any more with changes to pruning related to block merkle. I need to update our doc. This will also be a good opportunity to open source the document describing block merkle and deletes. Will work on that today along with the delete code.